token-lending/program/src/processor.rs

```diff
@@ -1176,6 +1184,363 @@ fn process_accrue_interest(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
     Ok(())
 }
 
+#[inline(never)] // avoid stack frame limit
+fn process_deposit_obligation_collateral(
+    program_id: &Pubkey,
+    collateral_amount: u64,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    if collateral_amount == 0 {
+        return Err(LendingError::InvalidAmount.into());
+    }
+
+    let account_info_iter = &mut accounts.iter();
+    let source_collateral_info = next_account_info(account_info_iter)?;
+    let deposit_reserve_info = next_account_info(account_info_iter)?;
+    let deposit_reserve_collateral_supply_info = next_account_info(account_info_iter)?;
+    let borrow_reserve_info = next_account_info(account_info_iter)?;
+    let obligation_info = next_account_info(account_info_iter)?;
+    let obligation_token_mint_info = next_account_info(account_info_iter)?;
+    let obligation_token_output_info = next_account_info(account_info_iter)?;
+    let lending_market_info = next_account_info(account_info_iter)?;
+    let lending_market_authority_info = next_account_info(account_info_iter)?;
+    let user_transfer_authority_info = next_account_info(account_info_iter)?;
+    let memory = next_account_info(account_info_iter)?;
+    let clock = &Clock::from_account_info(next_account_info(account_info_iter)?)?;
+    let token_program_id = next_account_info(account_info_iter)?;
+
+    // Ensure memory is owned by this program so that we don't have to zero it out
+    if memory.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+
+    let lending_market = LendingMarket::unpack(&lending_market_info.data.borrow())?;
+    if lending_market_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &lending_market.token_program_id != token_program_id.key {
+        return Err(LendingError::InvalidTokenProgram.into());
+    }
+
+    let deposit_reserve = Reserve::unpack(&deposit_reserve_info.data.borrow())?;
+    // @TODO: check if this logic is valid for withdraw
+    if deposit_reserve_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    // @TODO: check if & reference is valid
+    if &deposit_reserve.lending_market != lending_market_info.key {
```

Comment:
I think what I was missing conceptually is how the shared borrow operator in Rust works.

<img width="595" alt="Screen Shot 2021-02-14 at 11 41 35 AM" src="https://user-images.githubusercontent.com/1173161/107884258-a4ee0700-6eb9-11eb-9234-1cddf96ce4ba.png">

This suggests `pubkey1` and `pubkey2` are equivalent (but not equal), and `pubkeyRef1` and `pubkeyRef2` are equal.

Basically, I was thinking that `&deposit_reserve.lending_market` might be equivalent to `(&deposit_reserve).lending_market`, but instead it's `&(deposit_reserve.lending_market)`, so what you said makes sense.

Reply:
Yup, looks like you got it. Worth pointing out that `Pubkey` implements the `Copy` trait so in the first two lines, you're copying the supply pubkey twice, once for `pubkey1` and once for `pubkey2`. If `Pubkey` didn't implement `Copy`, Rust would _move_ the value out of the collateral struct. And the compiler would complain because you can only do that once.